rest/route/pod_agent.go

```diff
@@ -328,3 +348,155 @@ func (h *podAgentNextTask) findDispatcher() (*dispatcher.PodDispatcher, error) {
+			"build_revision":          evergreen.BuildRevision,
+			"build_agent":             evergreen.AgentVersion,
+			"task_details_from_agent": h.details,
+			"pod_id":                  p.ID,
+		})
+		endTaskResp.ShouldExit = true
+		return gimlet.NewJSONResponse(endTaskResp)
+	}
+
+	// Clear the running task on the pod now that the task has finished.
+	if err := p.ClearRunningTask(); err != nil {
+		return gimlet.MakeJSONErrorResponder(errors.Wrapf(err, "clearing running task %s for pod %s", t.Id, p.ID))
+	}
+
+	projectRef, err := model.FindMergedProjectRef(t.Project, t.Version, true)
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(gimlet.ErrorResponse{
+			StatusCode: http.StatusInternalServerError,
+			Message:    errors.Wrap(err, "finding project").Error(),
+		})
+	}
+	if projectRef == nil {
+		return gimlet.MakeJSONErrorResponder(gimlet.ErrorResponse{
+			StatusCode: http.StatusNotFound,
+			Message:    errors.Errorf("project '%s' not found", t.Project).Error(),
+		})
+	}
+
+	// mark task as finished
+	deactivatePrevious := utility.FromBoolPtr(projectRef.DeactivatePrevious)
+	err = model.MarkEnd(t, evergreen.APIServerTaskActivator, finishTime, &h.details, deactivatePrevious)
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(errors.Wrapf(err, "calling mark finish on task %v", t.Id))
+	}
+
+	if t.Requester == evergreen.MergeTestRequester {
+		if err = model.HandleEndTaskForCommitQueueTask(t, h.details.Status); err != nil {
+			return gimlet.MakeJSONErrorResponder(err)
+		}
+	}
+
+	// the task was aborted if it is still in undispatched.
+	// the active state should be inactive.
+	if h.details.Status == evergreen.TaskUndispatched {
+		if t.Activated {
+			grip.Warningf("task %v is active and undispatched after being marked as finished", t.Id)
+			return gimlet.NewJSONResponse(&apimodels.EndTaskResponse{})
+		}
+		grip.Infof(fmt.Sprintf("task %v has been aborted and will not run", t.Id))
+		return gimlet.NewJSONResponse(&apimodels.EndTaskResponse{})
+	}
+
+	// GetDisplayTask will set the DisplayTask on t if applicable
+	// we set this before the collect task end job is run to prevent data race
+	dt, err := t.GetDisplayTask()
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(errors.Wrap(err, "getting display task"))
+	}
+	queue := h.env.RemoteQueue()
+	job := units.NewCollectTaskEndDataJob(t, nil, p)
+	if err = queue.Put(ctx, job); err != nil {
+		return gimlet.MakeJSONErrorResponder(errors.Wrap(err, "couldn't queue job to update task stats accounting"))
+	}
+
+	if p.Status != pod.StatusRunning {
+		endTaskResp.ShouldExit = true
+	}
+
+	msg := message.Fields{
+		"message":     "Successfully marked task as finished",
+		"task_id":     t.Id,
+		"execution":   t.Execution,
+		"operation":   "mark end",
+		"duration":    time.Since(finishTime),
+		"should_exit": endTaskResp.ShouldExit,
+		"status":      h.details.Status,
+	}
+
+	if dt != nil {
+		msg["display_task_id"] = t.DisplayTask.Id
```

Comment:
Would this be different from just `t.DisplayTaskId`?

Reply:
I still have this question. It seems like we don't need `t.GetDisplayTask` at all because the parent display task is never used, and we could instead use `t.DisplayTaskId` here.

Reply:
We need to call `GetDisplayTask` because it actually sets the display task. If we just call `t.DisplayTaskId` it may be nil

Reply:
I've changed this to now check `if t.IsPartOfDisplay()` at the end of the function which will call `GetDisplayTask` behind the scenes